File: weekday.hs
import Data.Time
import Data.Time.Calendar.WeekDate (toWeekDate)

data Period = Autumn
            | Winter
            | Spring
            deriving (Show, Eq)

period :: (Ord a, Num t) => (MonthOfYear -> t -> a) -> a -> Period
period thisYear today
    | today < thisYear February 9 = Winter
    | today >= thisYear September 1 = Autumn
    | otherwise = Spring 

week :: Period -> Day -> String
week Winter _ = "Хороших праздников, удачной сессии!"
week p d
    | dayOfWeek d == Sunday = "Сегодня воскресенье, лучше иди домой"
    | otherwise = show x ++ " неделя" where
                        x = 1 + x0 - x1 - if limitIsSunday then 1 else 0
                        limitIsSunday = dayOfWeek limit == Sunday 
                        (_, x1, _) = toWeekDate limit
                        (y, x0, _) = toWeekDate d
                        limit = if p == Spring
                            then this February 9
                            else this September 1
                        this = fromGregorian y
main :: IO ()
main = do
    now <- getCurrentTime
    let today = utctDay now
    --let today = fromGregorian 2024 September 9 -- it was TEST
    let (year,_,_) = toGregorian today
    putStr $ week (period (fromGregorian year) today) today

File: weekday.kt
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.Month

enum class Period {
    Autumn,
    Winter,
    Spring
}

fun determinePeriod(today: LocalDate): Period {
    val februaryCutoff = LocalDate.of(today.year, Month.FEBRUARY, 9)
    val septemberCutoff = LocalDate.of(today.year, Month.SEPTEMBER, 1)

    return when {
        today.isBefore(februaryCutoff) -> Period.Winter
        today.isAfter(septemberCutoff.minusDays(1)) -> Period.Autumn
        else -> Period.Spring
    }
}

fun calculateWeek(period: Period, date: LocalDate): String {
    return when {
        period == Period.Winter -> "Хороших праздников, удачной сессии!"
        date.dayOfWeek == DayOfWeek.SUNDAY -> "Сегодня воскресенье, лучше иди домой"
        else -> {
            val currentWeek = date.getWeekOfYear()
            val periodLimit = when (period) {
                Period.Spring -> LocalDate.of(date.year, Month.FEBRUARY, 9)
                Period.Autumn -> LocalDate.of(date.year, Month.SEPTEMBER, 1)
                else -> throw IllegalStateException("Unexpected period")
            }
            val limitWeek = periodLimit.getWeekOfYear()
            val isLimitSunday = periodLimit.dayOfWeek == DayOfWeek.SUNDAY

            val weekOffset = if (isLimitSunday) 1 else 0
            val weekNumber = 1 + currentWeek - limitWeek - weekOffset
            "$weekNumber неделя"
        }
    }
}

fun LocalDate.getWeekOfYear(): Int {
    return this.get(java.time.temporal.ChronoField.ALIGNED_WEEK_OF_YEAR)
}

fun main() {
    val today = LocalDate.now()
    val period = determinePeriod(today)
    val weekInfo = calculateWeek(period, today)
    println(weekInfo)
}
